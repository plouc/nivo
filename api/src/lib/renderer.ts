import { pick } from 'lodash'
import React, { FunctionComponent } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { ChartProps, chartsMapping, ChartType } from '../mapping'

const staticProps = {
    animate: false,
    isInteractive: false,
    renderWrapper: false,
    theme: {},
}

export const renderChart = <T extends ChartType>(
    {
        type,
        props,
    }: {
        type: T
        props: ChartProps<T>
    },
    override
) => {
    const chart = chartsMapping[type]
    const component = chart.component as FunctionComponent<ChartProps<T>>
    const mergedProps = {
        ...staticProps,
        ...chart.defaults,
        ...props,
        ...pick(override, chart.runtimeProps || []),
    }
    const rendered = renderToStaticMarkup(
        React.createElement<ChartProps<T>>(component, mergedProps)
    )

    return `<?xml version="1.0" ?>${rendered}`
}
