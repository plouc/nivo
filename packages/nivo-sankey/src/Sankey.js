/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React from 'react'
import cloneDeep from 'lodash/cloneDeep'
import uniq from 'lodash/uniq'
import { sankey as d3Sankey } from 'd3-sankey'
import { Container, SvgWrapper } from '@nivo/core'
import { BoxLegendSvg } from '@nivo/legends'
import SankeyNodes from './SankeyNodes'
import SankeyLinks from './SankeyLinks'
import SankeyLabels from './SankeyLabels'
import { SankeyPropTypes, sankeyAlignmentFromProp } from './props'
import enhance from './enhance'

const getId = d => d.id

const Sankey = ({
    data: _data,

    align,

    // dimensions
    margin,
    width,
    height,
    outerWidth,
    outerHeight,

    // nodes
    nodeOpacity,
    nodeHoverOpacity,
    nodeHoverOthersOpacity,
    nodeWidth,
    nodePaddingX,
    nodePaddingY,
    nodeBorderWidth,
    getNodeBorderColor, // computed
    setCurrentNode, // injected
    currentNode, // injected

    // links
    linkOpacity,
    linkHoverOpacity,
    linkHoverOthersOpacity,
    linkContract,
    linkBlendMode,
    enableLinkGradient,
    getLinkColor, // computed
    setCurrentLink, // injected
    currentLink, // injected

    // labels
    enableLabels,
    getLabel,
    labelPosition,
    labelPadding,
    labelOrientation,
    getLabelTextColor, // computed

    // theming
    theme,
    getColor, // computed

    // tooltip
    nodeTooltip,
    linkTooltip,

    // motion
    animate,
    motionDamping,
    motionStiffness,

    // interactivity
    isInteractive,
    onClick,
    tooltipFormat,

    legends,
}) => {
    const sankey = d3Sankey()
        .nodeAlign(sankeyAlignmentFromProp(align))
        .nodeWidth(nodeWidth)
        .nodePadding(nodePaddingY)
        .size([width, height])
        .nodeId(getId)

    // deep clone is required as the sankey diagram mutates data
    const data = cloneDeep(_data)
    sankey(data)

    data.nodes.forEach(node => {
        node.color = getColor(node)
        node.label = getLabel(node)
        node.x = node.x0 + nodePaddingX
        node.y = node.y0
        node.width = Math.max(node.x1 - node.x0 - nodePaddingX * 2, 0)
        node.height = Math.max(node.y1 - node.y0, 0)
    })

    data.links.forEach(link => {
        link.color = getLinkColor(link)
    })

    const legendData = data.nodes.map(node => ({
        label: node.label,
        fill: node.color,
    }))

    const motionProps = {
        animate,
        motionDamping,
        motionStiffness,
    }

    let isCurrentNode = () => false
    let isCurrentLink = () => false

    if (currentLink) {
        isCurrentNode = ({ id }) => id === currentLink.source.id || id === currentLink.target.id
        isCurrentLink = ({ source, target }) =>
            source.id === currentLink.source.id && target.id === currentLink.target.id
    }

    if (currentNode) {
        let currentNodeIds = [currentNode.id]
        data.links
            .filter(
                ({ source, target }) => source.id === currentNode.id || target.id === currentNode.id
            )
            .forEach(({ source, target }) => {
                currentNodeIds.push(source.id)
                currentNodeIds.push(target.id)
            })

        currentNodeIds = uniq(currentNodeIds)
        isCurrentNode = ({ id }) => currentNodeIds.includes(id)
        isCurrentLink = ({ source, target }) =>
            source.id === currentNode.id || target.id === currentNode.id
    }

    return (
        <Container isInteractive={isInteractive} theme={theme}>
            {({ showTooltip, hideTooltip }) => (
                <SvgWrapper width={outerWidth} height={outerHeight} margin={margin}>
                    <SankeyLinks
                        links={data.links}
                        linkContract={linkContract}
                        linkOpacity={linkOpacity}
                        linkHoverOpacity={linkHoverOpacity}
                        linkHoverOthersOpacity={linkHoverOthersOpacity}
                        linkBlendMode={linkBlendMode}
                        enableLinkGradient={enableLinkGradient}
                        showTooltip={showTooltip}
                        hideTooltip={hideTooltip}
                        setCurrentLink={setCurrentLink}
                        currentNode={currentNode}
                        currentLink={currentLink}
                        isCurrentLink={isCurrentLink}
                        onClick={onClick}
                        tooltip={linkTooltip}
                        theme={theme}
                        tooltipFormat={tooltipFormat}
                        {...motionProps}
                    />
                    <SankeyNodes
                        nodes={data.nodes}
                        nodePaddingX={nodePaddingX}
                        nodeOpacity={nodeOpacity}
                        nodeHoverOpacity={nodeHoverOpacity}
                        nodeHoverOthersOpacity={nodeHoverOthersOpacity}
                        nodeBorderWidth={nodeBorderWidth}
                        getNodeBorderColor={getNodeBorderColor}
                        showTooltip={showTooltip}
                        hideTooltip={hideTooltip}
                        setCurrentNode={setCurrentNode}
                        currentNode={currentNode}
                        currentLink={currentLink}
                        isCurrentNode={isCurrentNode}
                        onClick={onClick}
                        tooltip={nodeTooltip}
                        theme={theme}
                        tooltipFormat={tooltipFormat}
                        {...motionProps}
                    />
                    {enableLabels && (
                        <SankeyLabels
                            nodes={data.nodes}
                            width={width}
                            labelPosition={labelPosition}
                            labelPadding={labelPadding}
                            labelOrientation={labelOrientation}
                            getLabelTextColor={getLabelTextColor}
                            theme={theme}
                            {...motionProps}
                        />
                    )}
                    {legends.map((legend, i) => (
                        <BoxLegendSvg
                            key={i}
                            {...legend}
                            containerWidth={width}
                            containerHeight={height}
                            data={legendData}
                        />
                    ))}
                </SvgWrapper>
            )}
        </Container>
    )
}

Sankey.propTypes = SankeyPropTypes

export default enhance(Sankey)
