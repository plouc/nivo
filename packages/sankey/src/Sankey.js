/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React from 'react'
import { uniq } from 'lodash'
import { Container, SvgWrapper } from '@nivo/core'
import { BoxLegendSvg } from '@nivo/legends'
import { SankeyPropTypes } from './props'
import enhance from './enhance'
import SankeyNodes from './SankeyNodes'
import SankeyLinks from './SankeyLinks'
import SankeyLabels from './SankeyLabels'

const Sankey = ({
    nodes,
    links,
    layout,

    margin,
    width,
    height,
    outerWidth,
    outerHeight,

    nodeOpacity,
    nodeHoverOpacity,
    nodeHoverOthersOpacity,
    nodeBorderWidth,
    getNodeBorderColor, // computed
    setCurrentNode, // injected
    currentNode, // injected

    linkOpacity,
    linkHoverOpacity,
    linkHoverOthersOpacity,
    linkContract,
    linkBlendMode,
    enableLinkGradient,
    setCurrentLink, // injected
    currentLink, // injected

    enableLabels,
    labelPosition,
    labelPadding,
    labelOrientation,
    getLabelTextColor, // computed

    theme,

    nodeTooltip,
    linkTooltip,

    animate,
    motionDamping,
    motionStiffness,

    isInteractive,
    onClick,
    tooltipFormat,

    legends,
    legendData,
}) => {
    const motionProps = {
        animate,
        motionDamping,
        motionStiffness,
    }

    let isCurrentNode = () => false
    let isCurrentLink = () => false

    if (currentLink) {
        isCurrentNode = ({ id }) => id === currentLink.source.id || id === currentLink.target.id
        isCurrentLink = ({ source, target }) =>
            source.id === currentLink.source.id && target.id === currentLink.target.id
    }

    if (currentNode) {
        let currentNodeIds = [currentNode.id]
        links
            .filter(
                ({ source, target }) => source.id === currentNode.id || target.id === currentNode.id
            )
            .forEach(({ source, target }) => {
                currentNodeIds.push(source.id)
                currentNodeIds.push(target.id)
            })

        currentNodeIds = uniq(currentNodeIds)
        isCurrentNode = ({ id }) => currentNodeIds.includes(id)
        isCurrentLink = ({ source, target }) =>
            source.id === currentNode.id || target.id === currentNode.id
    }

    return (
        <Container isInteractive={isInteractive} theme={theme}>
            {({ showTooltip, hideTooltip }) => (
                <SvgWrapper width={outerWidth} height={outerHeight} margin={margin} theme={theme}>
                    <SankeyLinks
                        links={links}
                        layout={layout}
                        linkContract={linkContract}
                        linkOpacity={linkOpacity}
                        linkHoverOpacity={linkHoverOpacity}
                        linkHoverOthersOpacity={linkHoverOthersOpacity}
                        linkBlendMode={linkBlendMode}
                        enableLinkGradient={enableLinkGradient}
                        showTooltip={showTooltip}
                        hideTooltip={hideTooltip}
                        setCurrentLink={setCurrentLink}
                        currentNode={currentNode}
                        currentLink={currentLink}
                        isCurrentLink={isCurrentLink}
                        onClick={onClick}
                        tooltip={linkTooltip}
                        theme={theme}
                        tooltipFormat={tooltipFormat}
                        {...motionProps}
                    />
                    <SankeyNodes
                        nodes={nodes}
                        nodeOpacity={nodeOpacity}
                        nodeHoverOpacity={nodeHoverOpacity}
                        nodeHoverOthersOpacity={nodeHoverOthersOpacity}
                        nodeBorderWidth={nodeBorderWidth}
                        getNodeBorderColor={getNodeBorderColor}
                        showTooltip={showTooltip}
                        hideTooltip={hideTooltip}
                        setCurrentNode={setCurrentNode}
                        currentNode={currentNode}
                        currentLink={currentLink}
                        isCurrentNode={isCurrentNode}
                        onClick={onClick}
                        tooltip={nodeTooltip}
                        theme={theme}
                        tooltipFormat={tooltipFormat}
                        {...motionProps}
                    />
                    {enableLabels && (
                        <SankeyLabels
                            nodes={nodes}
                            layout={layout}
                            width={width}
                            height={height}
                            labelPosition={labelPosition}
                            labelPadding={labelPadding}
                            labelOrientation={labelOrientation}
                            getLabelTextColor={getLabelTextColor}
                            theme={theme}
                            {...motionProps}
                        />
                    )}
                    {legends.map((legend, i) => (
                        <BoxLegendSvg
                            key={i}
                            {...legend}
                            containerWidth={width}
                            containerHeight={height}
                            data={legendData}
                            theme={theme}
                        />
                    ))}
                </SvgWrapper>
            )}
        </Container>
    )
}

Sankey.propTypes = SankeyPropTypes

const enhancedSankey = enhance(Sankey)
enhancedSankey.displayName = 'Sankey'

export default enhancedSankey
