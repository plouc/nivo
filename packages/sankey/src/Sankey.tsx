import { createElement, Fragment, ReactNode } from 'react'
import uniq from 'lodash/uniq'
import { SvgWrapper, useDimensions, Container } from '@nivo/core'
import { BoxLegendSvg } from '@nivo/legends'
import { svgDefaultProps } from './props'
import { useSankey } from './hooks'
import { SankeyNodes } from './SankeyNodes'
import { SankeyLinks } from './SankeyLinks'
import { SankeyLabels } from './SankeyLabels'
import {
    DefaultLink,
    DefaultNode,
    SankeyLayerId,
    SankeyLinkDatum,
    SankeyNodeDatum,
    SankeySvgProps,
} from './types'

type InnerSankeyProps<N extends DefaultNode, L extends DefaultLink> = Omit<
    SankeySvgProps<N, L>,
    'animate' | 'motionConfig' | 'renderWrapper' | 'theme'
>

const InnerSankey = <N extends DefaultNode, L extends DefaultLink>({
    data,
    valueFormat,
    layout = svgDefaultProps.layout,
    sort = svgDefaultProps.sort,
    align = svgDefaultProps.align,
    width,
    height,
    margin: partialMargin,
    colors = svgDefaultProps.colors,
    nodeThickness = svgDefaultProps.nodeThickness,
    nodeSpacing = svgDefaultProps.nodeThickness,
    nodeInnerPadding = svgDefaultProps.nodeInnerPadding,
    nodeBorderColor = svgDefaultProps.nodeBorderColor,
    nodeOpacity = svgDefaultProps.nodeOpacity,
    nodeHoverOpacity = svgDefaultProps.nodeHoverOpacity,
    nodeHoverOthersOpacity = svgDefaultProps.nodeHoverOthersOpacity,
    nodeBorderWidth = svgDefaultProps.nodeBorderWidth,
    nodeBorderRadius = svgDefaultProps.nodeBorderRadius,
    linkOpacity = svgDefaultProps.linkOpacity,
    linkHoverOpacity = svgDefaultProps.linkHoverOpacity,
    linkHoverOthersOpacity = svgDefaultProps.linkHoverOthersOpacity,
    linkContract = svgDefaultProps.linkContract,
    linkBlendMode = svgDefaultProps.linkBlendMode,
    enableLinkGradient = svgDefaultProps.enableLinkGradient,
    enableLabels = svgDefaultProps.enableLabels,
    labelPosition = svgDefaultProps.labelPosition,
    labelPadding = svgDefaultProps.labelPadding,
    labelOrientation = svgDefaultProps.labelOrientation,
    label = svgDefaultProps.label,
    labelTextColor = svgDefaultProps.labelTextColor,
    nodeTooltip = svgDefaultProps.nodeTooltip,
    linkTooltip = svgDefaultProps.linkTooltip,
    isInteractive = svgDefaultProps.isInteractive,
    onClick,
    legends = svgDefaultProps.legends,
    layers = svgDefaultProps.layers,
    role = svgDefaultProps.role,
    ariaLabel,
    ariaLabelledBy,
    ariaDescribedBy,
}: InnerSankeyProps<N, L>) => {
    const { margin, innerWidth, innerHeight, outerWidth, outerHeight } = useDimensions(
        width,
        height,
        partialMargin
    )

    const {
        nodes,
        links,
        legendData,
        getNodeBorderColor,
        currentNode,
        setCurrentNode,
        currentLink,
        setCurrentLink,
        getLabelTextColor,
    } = useSankey<N, L>({
        data,
        valueFormat,
        layout,
        width: innerWidth,
        height: innerHeight,
        sort,
        align,
        colors,
        nodeThickness,
        nodeSpacing,
        nodeInnerPadding,
        nodeBorderColor,
        label,
        labelTextColor,
    })

    let isCurrentNode: (node: SankeyNodeDatum<N, L>) => boolean = () => false
    let isCurrentLink: (link: SankeyLinkDatum<N, L>) => boolean = () => false

    if (currentLink) {
        isCurrentNode = ({ id }: SankeyNodeDatum<N, L>) =>
            id === currentLink.source.id || id === currentLink.target.id
        isCurrentLink = ({ source, target }: SankeyLinkDatum<N, L>) =>
            source.id === currentLink.source.id && target.id === currentLink.target.id
    }

    if (currentNode) {
        let currentNodeIds = [currentNode.id]
        links
            .filter(
                ({ source, target }) => source.id === currentNode.id || target.id === currentNode.id
            )
            .forEach(({ source, target }) => {
                currentNodeIds.push(source.id)
                currentNodeIds.push(target.id)
            })

        currentNodeIds = uniq(currentNodeIds)
        isCurrentNode = ({ id }) => currentNodeIds.includes(id)
        isCurrentLink = ({ source, target }) =>
            source.id === currentNode.id || target.id === currentNode.id
    }

    const layerProps = {
        links,
        nodes,
        margin,
        width,
        height,
        outerWidth,
        outerHeight,
    }

    const layerById: Record<SankeyLayerId, ReactNode> = {
        links: null,
        nodes: null,
        labels: null,
        legends: null,
    }

    if (layers.includes('links')) {
        layerById.links = (
            <SankeyLinks<N, L>
                key="links"
                links={links}
                layout={layout}
                linkContract={linkContract}
                linkOpacity={linkOpacity}
                linkHoverOpacity={linkHoverOpacity}
                linkHoverOthersOpacity={linkHoverOthersOpacity}
                linkBlendMode={linkBlendMode}
                enableLinkGradient={enableLinkGradient}
                setCurrentLink={setCurrentLink}
                currentNode={currentNode}
                currentLink={currentLink}
                isCurrentLink={isCurrentLink}
                isInteractive={isInteractive}
                onClick={onClick}
                tooltip={linkTooltip}
            />
        )
    }

    if (layers.includes('nodes')) {
        layerById.nodes = (
            <SankeyNodes<N, L>
                key="nodes"
                nodes={nodes}
                nodeOpacity={nodeOpacity}
                nodeHoverOpacity={nodeHoverOpacity}
                nodeHoverOthersOpacity={nodeHoverOthersOpacity}
                borderWidth={nodeBorderWidth}
                borderRadius={nodeBorderRadius}
                getBorderColor={getNodeBorderColor}
                setCurrentNode={setCurrentNode}
                currentNode={currentNode}
                currentLink={currentLink}
                isCurrentNode={isCurrentNode}
                isInteractive={isInteractive}
                onClick={onClick}
                tooltip={nodeTooltip}
            />
        )
    }

    if (layers.includes('labels') && enableLabels) {
        layerById.labels = (
            <SankeyLabels<N, L>
                key="labels"
                nodes={nodes}
                layout={layout}
                width={innerWidth}
                height={innerHeight}
                labelPosition={labelPosition}
                labelPadding={labelPadding}
                labelOrientation={labelOrientation}
                getLabelTextColor={getLabelTextColor}
            />
        )
    }

    if (layers.includes('legends')) {
        layerById.legends = (
            <Fragment key="legends">
                {legends.map((legend, i) => (
                    <BoxLegendSvg
                        key={`legend${i}`}
                        {...legend}
                        containerWidth={innerWidth}
                        containerHeight={innerHeight}
                        data={legendData}
                    />
                ))}
            </Fragment>
        )
    }

    return (
        <SvgWrapper
            width={outerWidth}
            height={outerHeight}
            margin={margin}
            role={role}
            ariaLabel={ariaLabel}
            ariaLabelledBy={ariaLabelledBy}
            ariaDescribedBy={ariaDescribedBy}
        >
            {layers.map((layer, i) => {
                if (typeof layer === 'function') {
                    return <Fragment key={i}>{createElement(layer, layerProps)}</Fragment>
                }

                return layerById?.[layer] ?? null
            })}
        </SvgWrapper>
    )
}

export const Sankey = <N extends DefaultNode = DefaultNode, L extends DefaultLink = DefaultLink>({
    isInteractive = svgDefaultProps.isInteractive,
    animate = svgDefaultProps.animate,
    motionConfig = svgDefaultProps.motionConfig,
    theme,
    renderWrapper,
    ...otherProps
}: SankeySvgProps<N, L>) => (
    <Container
        {...{
            animate,
            isInteractive,
            motionConfig,
            renderWrapper,
            theme,
        }}
    >
        <InnerSankey<N, L> isInteractive={isInteractive} {...otherProps} />
    </Container>
)
