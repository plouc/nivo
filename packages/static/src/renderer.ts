import pick from 'lodash/pick'
import { createElement } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { ChartProps, chartsMapping, ChartType } from './mappings'

const staticProps = {
    animate: false,
    isInteractive: false,
    renderWrapper: false,
    theme: {},
}

export const renderChart = <T extends ChartType>(
    {
        type,
        props,
    }: {
        type: T
        props: ChartProps<T>
    },
    override: Partial<ChartProps<T>>
) => {
    const chart = chartsMapping[type]
    const component = chart.component
    const mergedProps = {
        ...staticProps,
        ...chart.defaults,
        ...props,
        ...pick(override, chart.runtimeProps || []),
    }
    const rendered = renderToStaticMarkup(
        // @ts-ignore
        createElement(component, mergedProps)
    )

    return `<?xml version="1.0" ?>${rendered}`
}
