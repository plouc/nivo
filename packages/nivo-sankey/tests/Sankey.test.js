import React from 'react'
import renderer from 'react-test-renderer'
import Sankey from '../src/Sankey'

const sampleNodes = [
    { id: 'john' },
    { id: 'raph' },
    { id: 'jane' },
    { id: 'marcel' },
    { id: 'raoul' },
]
const sampleLinks = [
    {
        source: 'john',
        target: 'jane',
        value: 75,
    },
    {
        source: 'raph',
        target: 'jane',
        value: 25,
    },
    {
        source: 'jane',
        target: 'marcel',
        value: 50,
    },
    {
        source: 'jane',
        target: 'raoul',
        value: 50,
    },
]

it('should render a basic sankey diagram', () => {
    const component = renderer.create(
        <Sankey
            width={500}
            height={300}
            data={{
                nodes: sampleNodes,
                links: sampleLinks,
            }}
        />
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})

it('should allow to disable labels', () => {
    const component = renderer.create(
        <Sankey
            width={500}
            height={300}
            data={{
                nodes: sampleNodes,
                links: sampleLinks,
            }}
            enableLabels={false}
        />
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})
