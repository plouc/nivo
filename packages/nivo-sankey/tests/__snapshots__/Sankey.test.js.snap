// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow to disable labels 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <svg
    height={300}
    width={500}
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(0,0)"
    >
      <g>
        <path
          d="M12,108.00000000000011C128,108.00000000000011,128,111.64598621738885,244,111.64598621738885"
          fill="none"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(232, 193, 160)"
          strokeOpacity={0.25}
          strokeWidth={216}
        />
        <path
          d="M12,264.0000000000001C128,264.0000000000001,128,255.64598621738884,244,255.64598621738884"
          fill="none"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(244, 117, 96)"
          strokeOpacity={0.25}
          strokeWidth={72}
        />
        <path
          d="M256,75.64598621738885C372,75.64598621738885,372,72.00000000000014,488,72.00000000000014"
          fill="none"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(241, 225, 91)"
          strokeOpacity={0.25}
          strokeWidth={144}
        />
        <path
          d="M256,219.64598621738884C372,219.64598621738884,372,228.0000000000001,488,228.0000000000001"
          fill="none"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(241, 225, 91)"
          strokeOpacity={0.25}
          strokeWidth={144}
        />
      </g>
      <g>
        <rect
          fill="rgb(232,193,160)"
          fillOpacity={0.75}
          height={216.00000000000003}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(194, 161, 134)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={0}
          y={1.1013412404281553e-13}
        />
        <rect
          fill="rgb(244,117,96)"
          fillOpacity={0.75}
          height={71.99999999999986}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(204, 98, 80)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={0}
          y={228.00000000000014}
        />
        <rect
          fill="rgb(241,225,91)"
          fillOpacity={0.75}
          height={288.00000000000017}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(202, 188, 76)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={244}
          y={3.64598621738885}
        />
        <rect
          fill="rgb(232,168,56)"
          fillOpacity={0.75}
          height={143.99999999999997}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(194, 141, 47)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={488}
          y={1.4210854715202004e-13}
        />
        <rect
          fill="rgb(97,205,187)"
          fillOpacity={0.75}
          height={143.9999999999999}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(81, 172, 156)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={488}
          y={156.0000000000001}
        />
      </g>
    </g>
  </svg>
</div>
`;

exports[`should render a basic sankey diagram 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <svg
    height={300}
    width={500}
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(0,0)"
    >
      <g>
        <path
          d="M12,108.00000000000011C128,108.00000000000011,128,111.64598621738885,244,111.64598621738885"
          fill="none"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(232, 193, 160)"
          strokeOpacity={0.25}
          strokeWidth={216}
        />
        <path
          d="M12,264.0000000000001C128,264.0000000000001,128,255.64598621738884,244,255.64598621738884"
          fill="none"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(244, 117, 96)"
          strokeOpacity={0.25}
          strokeWidth={72}
        />
        <path
          d="M256,75.64598621738885C372,75.64598621738885,372,72.00000000000014,488,72.00000000000014"
          fill="none"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(241, 225, 91)"
          strokeOpacity={0.25}
          strokeWidth={144}
        />
        <path
          d="M256,219.64598621738884C372,219.64598621738884,372,228.0000000000001,488,228.0000000000001"
          fill="none"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(241, 225, 91)"
          strokeOpacity={0.25}
          strokeWidth={144}
        />
      </g>
      <g>
        <rect
          fill="rgb(232,193,160)"
          fillOpacity={0.75}
          height={216.00000000000003}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(194, 161, 134)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={0}
          y={1.1013412404281553e-13}
        />
        <rect
          fill="rgb(244,117,96)"
          fillOpacity={0.75}
          height={71.99999999999986}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(204, 98, 80)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={0}
          y={228.00000000000014}
        />
        <rect
          fill="rgb(241,225,91)"
          fillOpacity={0.75}
          height={288.00000000000017}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(202, 188, 76)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={244}
          y={3.64598621738885}
        />
        <rect
          fill="rgb(232,168,56)"
          fillOpacity={0.75}
          height={143.99999999999997}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(194, 141, 47)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={488}
          y={1.4210854715202004e-13}
        />
        <rect
          fill="rgb(97,205,187)"
          fillOpacity={0.75}
          height={143.9999999999999}
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          stroke="rgb(81, 172, 156)"
          strokeOpacity={0.75}
          strokeWidth={1}
          width={12}
          x={488}
          y={156.0000000000001}
        />
      </g>
      <g>
        <text
          alignmentBaseline="central"
          style={
            Object {
              "fill": "rgb(174,145,120)",
              "pointerEvents": "none",
            }
          }
          textAnchor="start"
          transform="translate(21, 108.00000000000013) rotate(0)"
        >
          john
        </text>
        <text
          alignmentBaseline="central"
          style={
            Object {
              "fill": "rgb(183,88,72)",
              "pointerEvents": "none",
            }
          }
          textAnchor="start"
          transform="translate(21, 264.00000000000006) rotate(0)"
        >
          raph
        </text>
        <text
          alignmentBaseline="central"
          style={
            Object {
              "fill": "rgb(181,169,68)",
              "pointerEvents": "none",
            }
          }
          textAnchor="start"
          transform="translate(265, 147.64598621738892) rotate(0)"
        >
          jane
        </text>
        <text
          alignmentBaseline="central"
          style={
            Object {
              "fill": "rgb(174,126,42)",
              "pointerEvents": "none",
            }
          }
          textAnchor="end"
          transform="translate(479, 72.00000000000013) rotate(0)"
        >
          marcel
        </text>
        <text
          alignmentBaseline="central"
          style={
            Object {
              "fill": "rgb(73,154,141)",
              "pointerEvents": "none",
            }
          }
          textAnchor="end"
          transform="translate(479, 228.00000000000006) rotate(0)"
        >
          raoul
        </text>
      </g>
    </g>
  </svg>
</div>
`;
